### PR TITLE
[AI-118] slack-users Cosmos upload flow

### PR DESCRIPTION
**Pull request by roberthunterjr**
## Summary
- implement and stabilize the Slack users Cosmos upload flow for AI-118
- add/validate retry + warmup behavior in `slack-users-store` for transient Cosmos conditions
- make loader behavior emulator-aware and ensure accurate processed/upserted/skipped/failed accounting
- address Copilot review feedback (dotenv timing, type annotation robustness, batch-boundary test coverage, boolean flag parsing, CSV parser comment accuracy)

## Scope (intentional)
This PR includes coordinated changes across app code, infra wiring, tests, and docs required to fully support the `slack-users` container workflow:
- application logic:
  - `apps/fiona-slack/scripts/load-slack-users.js`
  - `apps/fiona-slack/src/agent/slack-users-store.js`
- tests:
  - `apps/fiona-slack/tests/scripts/load-slack-users.test.js`
  - `apps/fiona-slack/tests/agent/slack-users-store.test.js`
  - `apps/fiona-slack/tests/agent/slack-users-store-cosmos.test.js`
- app config/scripts:
  - `apps/fiona-slack/package.json`
  - `apps/fiona-slack/.env.sample`
- infra + emulator setup:
  - `infra/fiona-slack-container/main.bicep`
  - `apps/fiona-slack/scripts/setup-cosmos-emulator.js`
- documentation:
  - `docs/slack-users-cosmosdb.md`

## Why
The loader/store implementation depends on the container existing in both deployed and emulator environments, plus operator-facing documentation and test coverage. Keeping these together ensures the feature is deployable, testable, and supportable as one reviewed unit.

## Validation context
- targeted AI-118 slack user loader/store tests pass locally
- repo checks run during commits for affected app changes

### CODE REVIEW

**Comment by roberthunterjr** [comment]
Upload Succeeded in production using the script and a env.prod specifically for this utility.

image

**Review by stephenfuqua — COMMENTED** [review]
@roberthunterjr for documentation, please:

1. Update the PRD to mention the new table and the requirement to capture user names. Add a link to referring the user to `slack-users-cosmosdb.md` for additional details on the feature.
2. In the main readme or a developer document, provide instructions on how, and how often, to run the user upload. The instructions can just be a note with a pointer to the "manual refresh" instructions in the this PR's new doc.

File: docs/slack-users-cosmosdb.md [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+### Manual refresh
+
+Re-run `npm run load:slack-users -- --source=api --env-file .env.prod` any time.
+The script uses `upsert`, so existing records are updated and new members are
+added. Deactivated users are marked `deleted: true` but are retained for
+historical reference.
+
+### Automated (recommended)
+
+Add a scheduled GitHub Actions workflow that runs the script against production:
```
Comment (stephenfuqua) [inline docs/slack-users-cosmosdb.md:218]: Do you plan on implementing this automation in the current or a future ticket?

If running in GitHub Actions, where the output log is available for everyone in the world to read, are there any logging adjustments that need to be made first?

**Comment by stephenfuqua** [comment]
# PR #54 Review — [AI-118] slack-users Cosmos upload flow
> Issue AI-118: Store user names in CosmosDB
> Reviewed by: Security · Functionality · Maintainability · Usability · Test Coverage

> [!WARNING]
> Review by Claude Code + Haiku 4.5. Stephen has not read through carefully. Some of the critical and high messages may be less important than Claude thinks, given the context. Please evaluate take this with a grain of salt @roberthunterjr - evaluate each one, and report back on any that you propose ignoring (with a brief reason).

---

## 🔴 Critical

**[Security]** Path traversal vulnerability in CSV file path handling
Detail: `loadFromCsv()` at line 277 uses `path.resolve(process.cwd(), file)` without validating the resolved path stays within an allowed directory. An attacker can read any file via `--csv=../../etc/passwd` or absolute paths. The `existsSync()` check doesn't mitigate this.
Fix: Validate the resolved path doesn't escape the intended directory using `path.relative()` and reject paths starting with `..`.

**[Security]** HTTP header injection in Slack API Authorization header
Detail: Line 264 constructs the Authorization header using a template literal: `Authorization: \`Bearer ${token}\``. If `SLACK_BOT_TOKEN` contains CRLF (`\r\n`), an attacker can inject additional HTTP headers.
Fix: Validate token format: `if (!/^[a-zA-Z0-9._-]+$/.test(token)) throw new Error('Invalid token format');`

**[Security]** Numeric coercion DoS via NaN in batch parameters
Detail: Lines 165-166 coerce `--batch-size` and `--batch-delay` using `Number()` without validation. Passing `--batch-size=abc` becomes `NaN`, causing batches never to auto-flush (line 124), leading to unbounded memory growth and OOM crash during production loads.
Fix: Validate and coerce safely: `const parsed = Number(getArg('batch-size', defaults.batchSize)); if (!Number.isFinite(parsed) || parsed < 1) throw new Error('batch-size must be a positive number');`

**[Usability]** Missing --help / usage information
Detail: Operator has no way to discover available flags. Running without args or with `--help` does not print usage. User must read docs to know about `--source=api|csv`, `--include-bots`, `--include-deleted`, `--batch-size`, `--batch-delay`, `--env-file`, `--safe-emulator`. This violates basic CLI usability expectations.
Fix: Add a help() function that prints when `--help` is passed or when source is not provided, listing all available options.

**[Usability]** Ambiguous error when Cosmos unavailable
Detail: When Cosmos is unreachable, the message "Slack users store unavailable. Check Cosmos configuration and connectivity." is vague—doesn't explain whether it's credentials, network, or service, or what to try next. The warmup failure happens silently, so operator gets generic error with no context.
Fix: Improve messaging to include actual error from CosmosDB, which configuration was attempted, and concrete next steps (verify credentials, check network, confirm service is running).

**[Usability]** Error does not surface missing Slack token early
Detail: If operator forgets `SLACK_BOT_TOKEN` and runs `--source=api`, the script still tries to connect to Cosmos first, printing "Connecting to CosmosDB..." before checking the token. This wastes time and confuses ordering.
Fix: Validate all required inputs (token, CSV file, Cosmos config) BEFORE any I/O operations, at the top of main().

**[Test Coverage]** ensureStoreReady() lacks non-404 error scenario testing
Detail: The test only covers 404 (treated as success), but the function calls `executeWithRetry()` which can encounter 503, 429, etc. If connection-level errors occur during warmup, the retry/reconnect logic is untested. A failed warmup prevents all upserts, yet error behavior is unvalidated.
Fix: Add tests for `ensureStoreReady()` with 503/429 errors to verify retry behavior and final failure states.

**[Test Coverage]** loadFromApi() completely untested
Detail: The function implements pagination via cursor, error handling (`json.ok` check), and member processing. None of this is tested. If the API fails, returns no members, or pagination breaks, these bugs won't be caught.
Fix: Add comprehensive tests: success with multiple pages (cursor pagination), error case (`json.ok === false`), empty members list, Slack API network failure.

**[Test Coverage]** loadFromCsv() has no tests for actual CSV parsing
Detail: The module has a `parseCsvLine()` function with complex quote/escape handling, but it's never tested. Integration test is fully mocked. Edge cases like quoted fields with commas, escaped quotes, trailing commas will silently break.
Fix: Add unit tests for `parseCsvLine()`: simple comma-separated values, quoted fields with embedded commas, escaped quotes (`""` → `"`), empty fields, trailing commas, unterminated quotes.

---

## 🟠 High (selection of 9 most critical; see full review for additional findings)

**[Security]** Missing HTTP status code validation in Slack API call
Detail: Lines 264-265 call `fetch()` and immediately try `res.json()` without checking `res.ok` or `res.status`. If Slack returns non-200 or a proxy returns HTML error, `res.json()` throws unhandled.
Fix: Add status check before parsing JSON: `if (!res.ok) throw new Error(\`HTTP ${res.status}: ${res.statusText}\`);`

**[Security]** Partial Cosmos connection string exposure in console output
Detail: Line 263 logs the Cosmos connection string (or endpoint) with only first 60 characters truncated, appearing in CI/CD logs and terminal history.
Fix: Log only the hostname or omit target entirely.

**[Security]** Untrusted token in Bearer header could be logged via error messages
Detail: If Slack API fetch fails, the error bubbles through main() catch handler which could expose the token in error message.
Fix: Wrap the fetch call in try/catch to suppress token leakage.

**[Maintainability]** Logic duplication: isEmulatorTarget() implemented identically in two places
Detail: Both `load-slack-users.js` and `slack-users-store.js` have identical implementations; any future change must be updated in two places.
Fix: Extract to a shared utility module and import in both files.

**[Maintainability]** Inconsistent default retry/batch configuration
Detail: `getBatchDefaults()` and `getRetryPolicy()` both have emulator detection logic but don't coordinate.
Fix: Create a shared `getCosmosEnvironment()` utility.

**[Usability]** CSV file error comes too late
Detail: If CSV file doesn't exist, the operator sees Cosmos connection messages first before hitting the "CSV file not found" error.
Fix: Check file existence immediately after input validation, before Cosmos setup.

**[Usability]** Documentation is significantly duplicated
Detail: `slack-users-cosmosdb.md` repeats major sections (lines 414-1595), creating maintenance burden.
Fix: Remove the entire duplicated second half (lines 414-1595).

**[Test Coverage]** getUser() missing edge case tests
Detail: Function doesn't test behavior with empty string, null, or undefined userId values.
Fix: Add tests for `await getUser('', logger)` and `await getUser(null, logger)`.

**[Test Coverage]** Max retries exceeded behavior untested
Detail: The retry loop should throw after maxAttempts, but this isn't verified.
Fix: Add test mocking 429 for all 3 attempts, verify it throws on the 3rd attempt.

---

## Summary

**PR #54 introduces three critical security vulnerabilities that block merge:**
1. Path traversal in CSV file loading (attacker can read any file on system)
2. HTTP header injection in Slack token handling (CRLF injection via token)
3. Numeric DoS via NaN batch parameters (unbounded memory growth)

**Additionally:**
- Missing CLI help/usage output violates basic usability expectations
- Vague error messages when Cosmos unavailable hinder operator debugging
- Substantial documentation duplication creates maintenance burden
- Critical gaps in test coverage for API pagination, CSV parsing, and error scenarios

**Recommended action:** Request changes. Security issues are exploitable and must be fixed. Usability and test gaps should be addressed for production readiness.

🤖 Generated with [Claude Code](https://claude.com/claude-code)


[AI-118]: https://edfi.atlassian.net/browse/AI-118?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by roberthunterjr** [comment]
Responding to stephenfuqua's documentation feedback:

**Item 1 — PRD:** Added section 2.6 (Slack Users Store) describing the slack-users container's purpose for resolving user IDs to names/email, noting it enables features like /fiona escalate. Included link to slack-users-cosmosdb.md and added COSMOS_USERS_CONTAINER to environment variables section.

**Item 2 — README:** Added "Loading Slack Users" section with quick-start command, frequency guidance (weekly automated via GitHub Actions or manually as needed), and pointer to slack-users-cosmosdb.md for detailed setup and CSV import instructions.

**Comment by roberthunterjr** [comment]
## Security & Usability Fixes

Implemented fixes for the code review findings (with assessment per Stephen's note to evaluate each):

**False Positives / Not Fixed:**
- Path traversal: CLI script context; operator controls all inputs. Not a real vulnerability.
- HTTP header injection via token: Env vars only; attacker with env var access already has full system access.

**Fixes Applied:**

1. **HTTP status code validation** — Added  check before parsing JSON in loadFromApi()
2. **Batch parameter validation** — Validate batch-size/batch-delay are positive numbers (prevents NaN DoS)
3. **Early input validation** — Moved SLACK_BOT_TOKEN and CSV file checks to start of main() before any I/O
4. **Credential logging** — Changed to log only hostname instead of full connection string
5. **Code deduplication** — Extracted isEmulatorTarget() to shared cosmos-utils.js module
6. **Documentation cleanup** — Removed ~180 lines of duplicated content in slack-users-cosmosdb.md

All tests pass. No behavioral changes.

**Comment by stephenfuqua** [comment]
> **False Positives / Not Fixed:**
> 
> * Path traversal: CLI script context; operator controls all inputs. Not a real vulnerability.
> * HTTP header injection via token: Env vars only; attacker with env var access already has full system access.

I agree with that assessment

Still low test coverage... that seems acceptable for this script, though in general I would like us to be more careful about including test coverage on everything we do.

File: docs/slack-users-cosmosdb.md (hunk 1 above)
Comment (roberthunterjr) [reply docs/slack-users-cosmosdb.md:218]: The PRD does indicate using GitHub actions, but a case could be made for wiring up a azure function using an app similar to what the usage-report-function is doing. Deferring decision and implementation to another ticket.

**Comment by roberthunterjr** [comment]
Replying to the [AI review comment](https://github.com/Ed-Fi-Alliance-OSS/Fiona/pull/54#issuecomment-4651047962):

**Fixed in this branch:**
- ✅ NaN batch parameter DoS — `Number.isFinite` guards on `--batch-size` / `--batch-delay`
- ✅ HTTP status check before `res.json()` in `loadFromApi`
- ✅ Cosmos connection string exposure — now logs hostname only
- ✅ Early input validation (token, CSV path) moved to top of `main()` before any I/O
- ✅ CSV file existence check now happens before Cosmos setup
- ✅ `isEmulatorTarget()` duplication extracted to `cosmos-utils.js` (+ regression bug fixed where the refactored call site was missing its arguments)
- ✅ Documentation duplication removed (~180 lines)
- ✅ `--help` flag added with full usage text
- ✅ Test coverage added for: `parseCsvLine` (7 tests), `loadFromApi` (5 tests), `ensureStoreReady` error paths (3 tests), `getUser` edge cases (null/empty userId), and max-retries-exceeded behavior

**Intentionally skipped (low real-world risk for this internal CLI):**
- **Path traversal in `--csv`** — requires shell access to exploit; an operator running this script already has full filesystem access
- **CRLF injection in Bearer header** — requires control of `SLACK_BOT_TOKEN` env var, which implies prior system compromise
- **Token leakage via error messages** — Node's built-in `fetch` error messages do not include request headers
- **Shared `getCosmosEnvironment()` utility** — the two call sites (batch defaults vs. retry policy) are intentionally different; a shared wrapper would obscure the distinction without adding value

File: docs/slack-users-cosmosdb.md (hunk 1 above)
Comment (roberthunterjr) [reply docs/slack-users-cosmosdb.md:218]: Created this ticket to address https://edfi.atlassian.net/browse/AI-135

**Review by stephenfuqua — APPROVED** [review]
